rgo/git.py

```diff
@@ -149,6 +149,52 @@ def describe(self, pkg=None):
         version = version.replace("-", "_")
         return version, release
 
+    def _check_output(self, cmd):
+        env = dict(os.environ)
+        env["LC_ALL"] = "C.UTF-8"
+        output = subprocess.check_output(cmd, cwd=self.cwd, encoding="utf-8", env=env).strip().splitlines()
+        return output
+
+    def get_describe_long(self, ref):
+        """
+        Run git describe --tags --long <ref>
+        Return the a tuple with the following data:
+          * desc: raw `git desribe --tags --long <ref>` output
+          * tag: last available tag in commits <= ref (parsed from desc)
+          * commits: number of commits between the tag and the ref (parsed from desc)
+          * hash: hash of the ref (parsed from desc)
+        """
+        cmd = ["git", "describe", "--tags", "--long", ref]
+        desc = self._check_output(cmd)[0].strip()
+
+        tag, commits, hash = desc.rsplit("-", 2)
+
+        # covert number of commits to integer
+        commits = int(commits)
+
+        # remove the 'g' prefix and keep only the hash
+        hash = hash[1:]
+
+        return desc, tag, commits, hash
+
+    def get_rpm_changelog(self):
+        """
+        Generate RPM changelog records from git history between self.ref and the last tag.
+        If self.ref == last tag, generate a changelog between self.ref and the last but one tag.
+        """
+        ref_desc, ref_tag, ref_commits, ref_hash = self.get_describe_long(self.ref)
+
+        if ref_commits == 0:
+            # ref is tagged, let's skip it and find another tag in the history
+            tag_desc, tag_tag, tag_commits, tag_hash = self.get_describe_long(ref_tag + "^")
```

Comment:
So I assume you do this to have "something" in the changelog. If the HEAD is tagged, get the list of changes from the previous tag instead of an empty changelog? If so, this is what should be described in the comment to clarify.

Reply:
TODO: explain why we're doing it